File: y2024_C#/AdventOfCode/Day_06/DaySix.cs
using AdventofCode._General;
using AdventOfCode._Utils;

namespace AdventOfCode.Day_06;

public class DaySix : Day {
    private readonly ObstacleMap _obstacleMap;
    private readonly Position _startPosition;
    private readonly Direction _startDirection;
    private readonly Walker _walker;

    public DaySix() : base("input_06.txt") {
        _obstacleMap = new ObstacleMap(Input);
        _startPosition = _obstacleMap.GetStartPosition();
        _startDirection = _obstacleMap.GetStartDirection();

        _walker = new Walker(_obstacleMap, _startPosition, _startDirection);
        _walker.Walk();
    }

    public override int PartOne() {
        WalkAndCountSteps();

        return ResultPartOne;
    }

    private void WalkAndCountSteps() {
        var steps = _walker.UniqueVisitedPositions;

        ResultPartOne = steps.Count;
    }

    public override int PartTwo() {
        CountPossibleObstacles();

        return ResultPartTwo;
    }

    private void CountPossibleObstacles() {
        var possibleObstaclePositions = _walker.UniqueVisitedPositions;
        var possibleObstaclesCount = possibleObstaclePositions
            .Select(possibleObstaclePosition => _obstacleMap.WithObstacleAt(possibleObstaclePosition))
            .Select(NewWalker)
            .Count(walker => walker.Loops());

        ResultPartTwo = possibleObstaclesCount;
    }

    private Walker NewWalker(ObstacleMap obstacleMap) {
        return new Walker(obstacleMap, _startPosition, _startDirection);
    }
}

File: y2025_Java/src/day06/DaySix.java
package day06;

import _general.Day;

public class DaySix extends Day {
    private final HomeworkSolver homeworkSolver;

    public DaySix() {
        this.homeworkSolver = new HomeworkSolver(input);
    }

    public static void main(String[] args) {
        var day = new DaySix();
        day.printResults();
    }

    @Override
    protected String partOne() {
        return homeworkSolver.solveGrandTotal();
    }

    @Override
    protected String partTwo() {
        return homeworkSolver.solveRightToLeftGrandTotal();
    }
}
